arm64: dts: qcom: xiaomi-tulip: enable touchscreen input and bluetooth

M0Rf30 opened the pull request:
As in the title

dmesg related lines

```
[25.060439] input: Novatek NT36XXX Touchscreen as /devices/platform/soc/c175000.i2c/i2c-0/0-0062/input/input2
```

related event device node is
`/dev/input/event2`

15/07
Hereby I also:
- add panel and descriptors in mainline driver
- add ncp and ldo voltage regulators as in downstream
- some little fixes to align with upstream naming
- dfps region
- disable dsi1 nodes (as in downstream, cause enabled in parent dtsi)
- remove non-working wifi and add only modem node.

22/07
- bluetooth

MarijnS95 committed:
gpu go 700mhz

Signed-off-by: Marijn Suijten <marijns95@gmail.com>
Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

kholk committed:
Input: Add Novatek NT36xxx touchscreen driver

This is a driver for the Novatek in-cell touch controller and
supports various chips from the NT36xxx family, currently
including NT36525, NT36672A, NT36676F, NT36772 and NT36870.

Functionality like wake gestures and firmware flashing is not
included: I am not aware of any of these DrIC+Touch combo
chips not including a non-volatile memory and it should be
highly unlikely to find one, since the touch firmware is
embedded into the DriverIC one, which is obviously necessary
to drive the display unit.

However, the necessary address for the firmware update
procedure was included into the address table in this driver
so, in the event that someone finds the need to implement it
for a reason or another, it will be pretty straightforward to.

This driver is lightly based on the downstream implementation [1].
[1] https://github.com/Rasenkai/caf-tsoft-Novatek-nt36xxx

Signed-off-by: AngeloGioacchino Del Regno <kholk11@gmail.com>

kholk committed:
dt-bindings: touchscreen: Add binding for Novatek NT36xxx series driver

Add binding for the Novatek NT36xxx series touchscreen driver.

Signed-off-by: AngeloGioacchino Del Regno <kholk11@gmail.com>
Reviewed-by: Krzysztof Kozlowski <krzk@kernel.org>
Reviewed-by: Rob Herring <robh@kernel.org>

Commit:
power: supply: introduce pmi8998 fuel guage driver

Commit:
power: supply: add battery charging status feature

Commit:
fg: clean and read charge full and max voltage from dts

nergzd723 committed:
arm64: dts: qcom: sdm660: Add initial support for xiaomi-clover

Initial support for Xiaomi Mi Pad 4.

Currently works:
* simple-framebuffer
* USB
* touchscreen
* GPIO buttons

Signed-off-by: Mark Hargreaves <clashclanacc2602@gmail.com>

Tooniis committed:
power: pmi8998_fg: Rename to qcom_fg and add support for PMI8994/6

This adds support for accessing SRAM. That includes requesting and releasing
access to SRAM by setting some registers and handling the mem-avail IRQ,
configuring access, and reading from/writing to it.

Pre-gen3 fuel gauges require reading from SRAM to get voltage, current
and temperature data. Getting capacity is identical to gen3.

Tested on Xiaomi Mi Note 2 with PMI8996.

Tooniis committed:
power: pmi8998_fg: Remove some trailing spaces and tabs

This should be squashed.

JamiKettunen committed:
power: supply: qcom_fg: Report online property

kholk committed:
iommu/arm-smmu-qcom: Allow choosing a custom bypass emulation context

It cannot be taken for granted that the last IOMMU context is free
and available for us to use it to emulate bypass streams and, at least
on MSM8998's lpass iommu, using the last one will produce a crash;
please note that this may not be only dependant on the SoC, but also
on the firmware version.

To overcome to this issue, allow specifying a different context for
bypass emulation with the optional DT property "qcom,bypass-cbndx":
if this property is not found this means that we are either booting
with ACPI instead or that we don't want to specify a custom cb because
the default one (the last context bank) is fine.

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>
(JAMI: fixup for v6.0-rc1)

kholk committed:
arm64: dts: sdm630: Override bypass emulation context for lpass, anoc2

On the LPASS and ANOC2 IOMMUs, we cannot use the last context to
emulate bypass streams: set the right context banks to avoid crashes.

kholk committed:
arm64: dts: qcom: sdm630: Disallow disabling secured iommu context banks

Some IOMMU context banks are secured and any attempt to disable them
during arm-smmu initialization will result in a system crash.
To work around this issue, add a list of context banks that will never
be disabled, but only reconfigured, at initialization time.

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>

kholk committed:
arm64: dts: qcom: Introduce SDM660 Xiaomi Mi 8 Lite (platina)

Add support for the SDM660 Xiaomi Platina (Mi 8 Lite).

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>

kholk committed:
iommu/arm-smmu: Allow skipping context bank disable at reset time

On some SoCs some IOMMU context banks are actively used from TZ
during system boot, or some hypervisor configurations will trigger
a system reset upon disabling some protected/secured CBs.

Allow skipping the disablement of such contexts at IOMMU reset time
during initialization with a new implementation detail to work around
this quirk.

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>

kholk committed:
iommu/arm-smmu-qcom: Avoid disabling secured context banks

Some Qualcomm SoCs' TZ/hypervisor configuration is disallowing the
disablement of some context banks, being them used for tzapps and/or
remote processors; any attempt to disable such CBs will result in
triggering a fault and the system will freeze and/or reset.

For this reason, get a list of context banks that should never get
disabled during smmu initialization through a DT array property
`qcom,reset-nodisable-cbs`.
It was chosen to not hardcode the CBs as this is dependant on the
SoC's firmware, which may vary on different boards.

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>
(JAMI: fixup for v6.0-rc1)

kholk committed:
iommu/arm-smmu-qcom: Don't modify sACR on hypervisor secured iommus

Avoid modifying the contents of the secure Auxiliary Control Register
on some Qualcomm SoCs: due to a hypervisor configuration on some
firmware versions, this would result in a system crash.

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>

kholk committed:
iommu/arm-smmu-qcom: Skip the TTBR1 quirk for MSM8998 and SDM630

Similarly to MSM8996 DragonBoard 820c, MSM8998 and SDM630 are equipped
with Adreno 5xx series, which doesn't have separate pagetables support
at the moment of writing.
Skip the TTBR1 quirk for these two SoCs as to get Adreno in a usable
state.

kholk committed:
iommu/arm-smmu-qcom: Add MSM8998 and SDM660 mss compatibles for identity

Add MSM8998 and SDM660's mss-pil compatibles to switch the default
iommu domain type to IDENTITY, as similarly required by SDM845 and
others.

Signed-off-by: AngeloGioacchino Del Regno <angelogioacchino.delregno@somainline.org>
(JAMI: fixup for 6.0-rc2)

kcxt committed:
dt-bindings: input: add Qualcomm SPMI haptics driver

Add bindings for qcom PMIC SPMI haptics driver.

Signed-off-by: Caleb Connolly <caleb@connolly.tech>

kcxt committed:
input: add Qualcomm SPMI haptics driver

Add support for the haptics found in pmi8998 and related PMICs.
Based on the ff-memless interface. Currently this driver provides
a partial implementation of hardware features.

This driver only supports LRAs (Linear Resonant Actuators) in the "buffer"
mode with a single wave pattern.

Signed-off-by: Caleb Connolly <caleb@connolly.tech>

MollySophia committed:
arm64: dts: qcom: sdm660-platina: Add several more features

* Add simple framebuffer display node and reserve memory for it.
* Enable qusb2phy0 and its vdd supplies
* Disable gpucc and mmcc, simplefb stops working after few seconds when these are enabled.
* Add node for touchscreen
* Fix volume keys

Signed-off-by: Molly Sophia <mollysophia379@gmail.com>

iAboothahir committed:
drm/panel-simple: Add NT36672_TXD FHD+ IPS LCD Panel

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

JamiKettunen committed:
ath10k: Fake MSA ready event after event server arrival

This finally allowed me to get past the 2 ath10k_info()'s in qmi.c!

However, this doesn't fix everything; it seems you *have* to run
diag-router from https://github.com/andersson/diag to calm the firmware
down and stop constantly crashing; need to look into a way to get Wi-Fi
working later without needing this debug crap to be running.

link: https://github.com/JamiKettunen/linux-mainline-oneplus5/commit/088eaa9153803e2b028e092f88539036442da4a3

Signed-off-by: Jami Kettunen <jami.kettunen@protonmail.com>

iAboothahir committed:
arm64: dts: qcom: Add initial support for asus-x00td

Initial support for Asus Zenfone Max Pro M1 (SDM636)
	* Currently works
	-Simple-framebuffer
	-eMMC,SDCard
	-Power Button
	-Volume-keys
        -Wifi(Partial)
        -Bluetooth
	-Touchscreen-NT36XXX
	-Display
	-GPU(Partial)
	-USB

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

Danct12 committed:
arm64: dts: qcom: sdm660-xiaomi-lavender: Add Touchscreen

The Xiaomi Redmi Note 7 uses a Novatek NT36672A touchscreen, so
add support for it.

Signed-off-by: Dang Huynh <danct12@riseup.net>

minlexx committed:
sdm660_defconfig

Generated with:

$ make savedefconfig && cp defconfig ../arch/arm64/configs/sdm660_defconfig

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

iAboothahir committed:
drm/msm: add missing a508/512 MODULE_FIRMWARE declarations

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: pm660: add charger node

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: pm660: Add fuel gauge The device use a fuel guage gen 3

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: pm660: add rradc node

Signed-off-by: Aboothahir U<aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: asus-x00td: Add simple-battery

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: asus-x00td: enable SMB2

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: asus-x00td: enable Fuel Gauge

Signed-off-by: iAboothahir <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: asus-x00td: enable rradc node

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

M0Rf30 committed:
arm64: dts: qcom: add device-tree for Redmi Note 6 Pro (tulip)

Commit 1: Initial support
Commit 2: enable touchscreen input and panel
Commit 3: Enable bluetooth

Add support for HCI wcn3990-bt on Xiaomi Redmi Note 6 Pro
tulip.

Signed-off-by: Gianluca Boiano <morf3089@gmail.com>

buddyjojo committed:
drm/panel: Add support for xiaomi-jasmine nt36672 panel

alexeymin: this needs to be integrated into panel-novatek-nt36672a.c
as variation with its own compatible string and own init sequences.

[1] https://elixir.bootlin.com/linux/latest/source/drivers/gpu/drm/panel/panel-novatek-nt36672a.c

Signed-off-by: Joe Mason <buddyjojo06@outlook.com>

buddyjojo committed:
arm64: dts: qcom: sdm660: Add device tree for Xiaomi Mi A2 (jasmine)

Add initial device tree for Xiaomi Mi A2 phone, codenam jasmine.

Supported features:
* UART serial output
* hw keys (volume up/down, power)
* regulators
* USB
* eMMC

Things lacking full upstream support:
* novatek touchscreen
* modem remote processor + wcn3390 Wi-Fi
* display, GPU, panel + backlight

Signed-off-by: Joe Mason <buddyjojo06@outlook.com>

iAboothahir committed:
arm64: dts: qcom: pm660: fix interrupts flag

fixes : 2c3c18df0c5076848b8e5a3153393f6c8065e166

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

iAboothahir committed:
arm64: dts: qcom: pm660: fix pm660_fg-interrupts flag

fixes : 98e1f707312642064f9fc0dfb3e90d85e5642d10

Signed-off-by: Aboothahir U <aboothahirpkd@gmail.com>

ElectrodeYT committed:
arm64: dts: qcom: sdm630: Add initial device tree for nokia-pl2

Signed-off-by: Alexander Richards <electrodeyt@gmail.com>

ElectrodeYT committed:
drm/panel Add support for Focaltech FT8716u panel

Signed-off-by: Alexander Richards <electrodeyt@gmail.com>

M0Rf30 commented:
@minlexx @Aboothahir I also noticed that other devices using the nt36525 use an unsupported `vcc-supply`

driver only supports optional power supplies named as vdd and vio (jasmine, lavender,  platina, ....)

minlexx commented:
> @minlexx @Aboothahir I also noticed that other devices using the nt36525 use an unsupported `vcc-supply`
> 
> driver only supports optional power supplies named as vdd and vio (jasmine, lavender, platina, ....)

Do you mean that other devices got it wrong or what?

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-xiaomi-tulip.dts:
```diff
@@ -518,6 +532,18 @@
+		pins = "gpio66";
+		drive-strength = <8>;
+		bias-pull-up;
+	};
```
`function = ` ? `gpio` ? can you link to your downstream source or where did you get that pinconf info from

M0Rf30 replied:
[tulip.dts.txt](https://github.com/sdm660-mainline/linux/files/11644907/tulip.dts.txt)
in the file above you'll find
```
			ts_mux {

				ts_active {
					phandle = <0x17c>;

					mux {
						pins = "gpio66\0gpio67";
						function = "gpio";
					};

					config {
						pins = "gpio66\0gpio67";
						drive-strength = <0x10>;
						bias-pull-up;
					};
				};

				ts_reset_suspend {
					phandle = <0x17e>;

					mux {
						pins = "gpio66";
						function = "gpio";
					};

					config {
						pins = "gpio66";
						drive-strength = <0x02>;
						bias-pull-down;
					};
				};

				ts_int_suspend {
					phandle = <0x17d>;

					mux {
						pins = "gpio67";
						function = "gpio";
					};

					config {
						pins = "gpio67";
						drive-strength = <0x02>;
						bias-disable;
					};
				};
			};
```

minlexx replied:
Then this definition should have `fuinction = "gpio";` otherwise it's no point of setting electrical properties of the pin. Overall conversion from downstream sould look like
```dts
	touchscreen@62 {
		...
		pinctrl-names = "default", "sleep";
		pinctrl-0 = <&ts_default>;
		pinctrl-1 = <&ts_int_sleep>, <&ts_reset_sleep>;
	};
...
...
&tlmm {
	ts_default: ts-pins-default-state {
		pins = "gpio66", "gpio67";
		function = "gpio";
		drive-strength = <16>;
		bias-pull-up;
	};

	ts_reset_sleep: ts-reset-sleep-state {
		pins = "gpio66";
		function = "gpio";
		drive-strength = <2>;
		bias-pull-down;
	};

	ts_int_sleep: ts-int-sleep-state {
		pins = "gpio67";
		function = "gpio";
		drive-strength = <2>;
		bias-disable;
	};
};
```

M0Rf30 replied:
Done and re-tested on the device

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-xiaomi-tulip.dts:
```diff
@@ -195,6 +195,20 @@
 &blsp_i2c1 {
 	status = "okay";
+
+	touchscreen: novatek@62 {
```
The node should be named "touchscreen" and label is not needed (you don't reference `&touchscreen` anywhere, do you?)

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/writing-bindings.rst#n24

> use node names matching the class of the device

minlexx replied:
```diff
-touchscreen: novatek@62 {
+touchscreen@62 {
```

M0Rf30 commented:
> > @minlexx @Aboothahir I also noticed that other devices using the nt36525 use an unsupported `vcc-supply`
> > driver only supports optional power supplies named as vdd and vio (jasmine, lavender, platina, ....)
> 
> Do you mean that other devices got it wrong or what?

Yes, they got it wrong as you can see here
https://github.com/sdm660-mainline/linux/blob/sdm660-next-stable/drivers/input/touchscreen/nt36xxx.c#L712

in their dmesg, right now they should have
```
nt36xxx_ts 0-0062: supply vdd not found, using dummy regulator
nt36xxx_ts 0-0062: supply vio not found, using dummy regulator
```
simply because they are using vcc-supply (instead of vdd or vio)

minlexx committed:
[FOR UPSTREAM] arm64: dts: qcom: sdm630: Add support for WCN3990 Wi-Fi

Describe Wi-Fi device node used in SDM630/660.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
[FOR UPSTREAM] arm64: dts: qcom: sdm660-lavender: Enable radio interfaces

Enable support for modem remote processor, WCN3990 Wi-Fi and bluetooth
on Xiaomi Redmi Note 7 (lavender).

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

M0Rf30 committed:
drm/panel: nt36672a: add xiaomi-tulip panel support

Signed-off-by: Gianluca Boiano <morf3089@gmail.com>

minlexx committed:
[UPSTREAM] arm64: dts: qcom: sdm630: Add support for modem remoteproc

Modem subsystem in SDM630/660 is similar to MSM8998 and
device tree node for it is based on the one from msm8998.dtsi.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
Link: https://lore.kernel.org/r/20230719093458.2668842-1-alexeymin@postmarketos.org
Signed-off-by: Bjorn Andersson <andersson@kernel.org>

krzk committed:
[UPSTREAM] arm64: dts: qcom: sdm630: align USB DWC3 clocks with bindings

Bindings require different order of clocks for USB DWC3 nodes (sleep
before mock_utmi).

Signed-off-by: Krzysztof Kozlowski <krzysztof.kozlowski@linaro.org>
Reviewed-by: Konrad Dybcio <konrad.dybcio@linaro.org>
Link: https://lore.kernel.org/r/20230723141849.93078-2-krzysztof.kozlowski@linaro.org
Signed-off-by: Bjorn Andersson <andersson@kernel.org>

lweiss-fairphone committed:
[UPSTREAM] arm64: dts: qcom: Use QCOM_SCM_VMID defines for qcom,vmid

Since we have those defines available in a header, let's use them
everywhere where qcom,vmid property is used.

Signed-off-by: Luca Weiss <luca.weiss@fairphone.com>
Reviewed-by: Konrad Dybcio <konrad.dybcio@linaro.org>
Reviewed-by: Bryan O'Donoghue <bryan.odonoghue@linaro.org>
Link: https://lore.kernel.org/r/20230818-qcom-vmid-defines-v1-1-45b610c96b13@fairphone.com
Signed-off-by: Bjorn Andersson <andersson@kernel.org>

minlexx committed:
[FOR UPSTREAM] arm64: dts: qcom: sdm630: Add modem metadata mem

Similarly to MSM8998, add and use modem metadata memory region.
This fixes dtbs check warning; but however does not seemingly
affect device functionality.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

minlexx committed:
[FOR_UPSTREAM] arm64: dts: qcom: sdm630: remove refs to nonexistent clocks

Since commit d6edc31 ("clk: qcom: smd-rpm: Separate out
interconnect bus clocks") sdm660 RPM clocks driver no longer
provides following clocks:

- RPM_SMD_SNOC_CLK, RPM_SMD_SNOC_A_CLK
- RPM_SMD_CNOC_CLK, RPM_SMD_CNOC_A_CLK
- RPM_SMD_BIMC_CLK, RPM_SMD_BIMC_A_CLK
- RPM_SMD_MMSSNOC_AXI_CLK, RPM_SMD_MMSSNOC_AXI_CLK_A
- RPM_SMD_AGGR2_NOC_CLK, RPM_SMD_AGGR2_NOC_A_CLK

Remove references to them from device tree to fix probe
failures and get devices booting again.

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>

M0Rf30 committed:
arm64: dts: qcom: xiaomi-tulip: enable touchscreen input and bluetooth

Signed-off-by: Gianluca Boiano <morf3089@gmail.com>

M0Rf30 commented:
@minlexx I've done: 
- general cleanup
- add resin (vol-)
- enable bluetooth
- enable touchscreen

nothing to do with wifi and panel enablement

M0Rf30 commented:
[dmesg-tulip-6.6.txt](https://github.com/sdm660-mainline/linux/files/13468033/dmesg-tulip-6.6.txt)

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-xiaomi-tulip.dts:
```diff
@@ -174,7 +110,7 @@
 	status = "okay";
 
-	touchscreen@62 {
+	touchscreen: novatek@62 {
```
This is still not resolved. I do not understand why you changed it back after I fixed this line in 6.6 branch for you.

To be clear on terminology, naming of things in device tree is:
```
node_label: node_name@address { ... };
```
Here, your node_label (`touchscreen`) is not needed because it is never referenced anywhere and only takes more space in resulting compiled DTB file.

`node_name` should describe generic class of device, in this case `novatek` is not a class of device, `touchscreen` is.

Please change this line back to how it was: remove unused label and change name to be generic, `touchscreen@62`

M0Rf30 replied:
got confused by lavender dts, where also a label is reported (I always check also in comparison with other DTSes)

minlexx replied:
🙈  I'm fixing this myself in 6.6-wip branch... https://github.com/sdm660-mainline/linux/commit/b4c33d266df108412aed9ae8040b497ca319bf95

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-xiaomi-tulip.dts:
```diff
@@ -631,24 +503,13 @@
 	firmware-name = "qcom/venus-4.4/venus.mdt";
-};
 
-&wifi {
```
Why did you remove wi-fi node? Even if it is not fully working, we can still have it here, I think.

I also think you're using kernel config which is different from mine, can I see your defconfig? (Run `make savedefconfig`and share defconfig file somewhere? To be precisefull command will look something like `make ARCH=arm64 CROSS_COMPILE=aarch64-alpine-linux-musl- savedefconfig`)

minlexx commented:
I've picked the commit https://github.com/M0Rf30/linux/commit/2ca7fcb334cabf4b4f974947a0e92d86366d36c8 to qcom-sdm660-6.6.y and squashed it with commit adding initial support for tulip. Since device dts is not upstream, it makes little sense to have separate millions commits that fix each other, because during upstreaming you'll need single commit (there should not be commits that change previous commits in single patchset scope). It now looks like this 5bbbd17630a89f89581fa1a1d9c75e192f91d201 . Likely more force pushes are incoming, since I see now that pm660_rradc and remoteproc_mss are specified twice :see_no_evil: ( fixup! ace6397b209cc05ecf5b36103382a856538be76c ) and I still need to do something with xiaomi-clover...

P.S. nt36672a panel support for tulip is still separate commit
